Add a modal to warn users this is a demo

maxime-rainville opened the pull request:
Displays a pretty modal telling the user this is a demo site.

When you click the "I understand" button in the modal this sets a cookie using front end JS.

The Page Controller will include some JS/CSS/template when that cookie is not set.

# Parent issue
* https://github.com/silverstripeltd/open-sourcerers/issues/68

# Related PR
* https://github.com/silverstripe/bambusa-theme/pull/5

Commit:
Add a modal to warn users this is a demo.

maxime-rainville commented on app/css/dialog.scss:
```diff
@@ -0,0 +1,74 @@
+// These styles are only use for the privacy modal. This is meant to look like sscom.
```
This file is being compiled down to CSS manually. It's the only scss files so I fugred there wasn't much point having an entire build system just for it.

maxime-rainville commented on app/js/dialog.js:
```diff
@@ -0,0 +1,8 @@
+jQuery(function($) {
+  $('#privacyModal')
+    .modal({backdrop: 'static', keyboard: false})
```
This line prevents you from dismissing the modal by clicking on the backdrop or by hitting the Esc key.

This forces you to click the "I understand button".

chillu replied:
Nice touch, great attention to detail there

chillu commented on app/js/dialog.js:
```diff
@@ -0,0 +1,8 @@
+    .modal({backdrop: 'static', keyboard: false})
+    .modal('show')
+    .on('hide.bs.modal', function () {
+      document.cookie = "privacy_modal_showed=1; path=/; domain=" + privacy_modal_hostname;
```
Why the explicit domain setting? This makes it harder to achieve https://github.com/silverstripe/bambusa-installer/pull/15, because I need to set the global via a PHP controller there as well. https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie#Write_a_new_cookie says that you don't need the domain - I'm assuming the domain from `document.location.host` would match `Director::host()`?

maxime-rainville replied:
I've removed that bit.

chillu commented on app/src/PageController.php:
```diff
@@ -3,9 +3,45 @@
+         * Content of the Modal window if our show modal is enabled.
+         * @return \SilverStripe\ORM\FieldType\DBHTMLText|void
+         */
+        public function ModalWindow()
```
So you've kept the name generic because that goes into the bambusa theme, but the specific implementation of a "privacy modal" is an application concern, right? it's a bit confusing, but I'd agree with that reasoning.

maxime-rainville replied:
Yeah that was the idea.

chillu commented on app/src/PageController.php:
```diff
@@ -3,9 +3,45 @@
+
+            // Inject our privacy modal assets into the page
+            $cookie = Cookie::get('privacy_modal_showed');
+            if (empty($cookie)) {
```
We'll probably add HTTP caching to this, so I'd prefer if that conditional was handled on the client, rather than creating uncacheable responses.

maxime-rainville replied:
It's gone, all client side now ... FYI this means the modal and related style and JS will be served on every page.

Commit:
Move all the show modal cookie logic client side.

chillu commented on app/css/dialog.scss:
```diff
@@ -0,0 +1,74 @@
+
+  .modal-body {
+    padding: 2rem 1rem;
+    font-family: Metropolis, 'Proxima Nova', Arial, sans-serif;
```
Is there a reason you didn't use `$font-family-sans-serif`? It's got different fallback fonts, but the first two are the same

chillu commented on app/templates/PrivacyModal.ss:
```diff
@@ -0,0 +1,14 @@
+            </div>
+            <div class="modal-body">
+                <h2 id="privacyModalLabel">Welcome to your personal demo</h2>
+                <p>The content in this demo is used to demonstrate the features of SilverStripe <abbr title="Content Management System">CMS</abbr>. Please do not enter any personal or sensitive information.</p>
```
Hah even with an `<abbr>`, look at that.

chillu reviewed (approved):
One niggle about `$font-family-sans-serif`, but that doesn't block merge

chillu committed:
Limit linter to app/src

We now have app specific JS and CSS, which breaks the linter
since it doesn't filter by file extension.

chillu commented:
Travis failed on running `phpcs` over a `css` file, which I've fixed with b6f5d17c0ac2c15cd16fc55eab496cfd8659287a

chillu commented:
Note that there's a bug in this PR around cookie paths which I'm fixing as part of a related PR: https://github.com/silverstripe/bambusa-installer/pull/15

chillu commented:
I've also merged this PR on red, which caused a follow up issue: https://github.com/silverstripe/bambusa-installer/issues/21
